### DISCUSSION TITLE
Frequency tolerance

### DISCUSSION THREAD

**Discussion by faragher** [markqvist/Reticulum#285] [Help & Questions] [answered]
When I manually set 915 MHz using a serial connection, it seems to provide a slightly out of spec return frequency. I send

`C0 01 36 89 CA DB DC C0`

(FEND CMD_Frequency 36 89 CA C0 (escaped to DB DC) FEND)

When I get a response I receive:

`C0 1 36 89 C9 CB C0`

which provides a frequency of 915 MHz - 245 Hz. I notice the stock tolerance is +/- 100 Hz, is this something I'm doing wrong in the interface or is this simply within tolerance for the radio?

**Comment by markqvist** [comment, marked as answer]
This is due to the frequency resolution of the modem. I can't remember the exact number, but it's in the SX1276 datasheet somewhere. It's nothing to worry about though, for any frequency you can come up with, the practical offset is small enough to be irellevant.

**Reply by faragher** [reply]
Okay, I'm going to up the acceptable variance for ValidateRadioState. It does transmit/receive perfectly.

To be clear, and I should be keeping track of these tweaks, but this is for a C# port of the RNode interface and eventually a Reticulum compatible library, not the Python implementation. So far it controls the radio wonderfully and the interface is about 95% complete (Reconnect_Port, Validate_Firmware, callbacks, and some config are all that's left for the Interface) but I think Reticulum itself is the hurdle, not the radio controller.

**Reply by markqvist** [reply]
Very cool, great work! That will be really useful.

Yes, Reticulum itself will probably be a bit more effort, but you gotta start somewhere ;)

**Reply by LaneaLucy** [reply]
Does it really must be c#? I would really really appreciate a c or c++ port...

**Reply by faragher** [reply]
Short answer: yes.

Long answer: I am not a C programmer and the differences are substantial. .Net/Mono is extremely widely supported and has a rich standardized ecosystem. Outside of the Reticulum Network Stack, it requires no non-standard packages or external dependencies that can't be installed with NuGet. At time of writing, System.IO.Ports is the only package required outside of stock .Net 6. 

Since it can run on Windows, Mac, Linux, and Android (and even iOS, although that's another kettle of fish) it's an excellent toe-hold for programmers on any platform. Given the proper licenses, you could even attempt to run it on game consoles, but given the sandboxing, I doubt it would function as intended.

While it's certainly *possible* to go from C# intermediate language to C programmatically, there are fundamental differences that need to be addressed by a knowledgeable programmer. I am not up on the garbage collection and memory allocation the stack would need to function properly and indefinitely, meaning any C/C++ code I made would be buggy and have a huge memory footprint, the exact opposite of the requirements of embedded systems, which is an important use case of pure C. 

I do apologize if you're a skilled C programmer, you know all of this already, but I cannot make a C version of the stack that would be worth using. If you're interested in porting it to C, a large amount of effort has gone into identifying, typing, and declaring variables, something Python really glosses over, so it should be a real time saver starting with an implementation with explicit variable declaration.

Since Python and C# run on basically everything, the niche C fills is embedded systems, and that isn't something to be developed by someone who has almost no experience with pointers and malloc. Is it important for the ecosystem? Absolutely. Am I the one to do it? Absolutely not. When the C# implementation is complete, I'll have a better grasp of the stack, but learning the ins and outs of embedded system development would take more time than the port itself. Maybe in a few years.

**Reply by markqvist** [reply]
For what it's worth, I am close to starting a microcontroller-oriented C implementation.

In the grand scheme of things, we are "*this* close" to Reticulum being API-complete, and when that is done, I will start actively working on the C implementation, as soon as time and resources allow it.

**Reply by LaneaLucy** [reply]
Microcontroller is the scenario why I wanted a c port. So do you have a rough time estimate for when the first version of it can be tested? And I don't talk about a release, I talk about Dev versions. Main interest for me would be rns on esp with network (WiFi) interface and lora interface to have a small, low power repeater
